doc/man/pilot-install-hinote.1
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "pilot-install-hino" 1 "Copyright 1996-2007 FSF" "0.12.5" "PILOT-LINK"
.SH NAME
pilot-install-hinote \- Install local files into your Hi-Note database on your Palm handheld.
.SH "SECTION"

.PP
pilot\-link: Userland conduits

.SH "SYNOPSIS"

.PP
 \fIpilot\-install\-hinote\fR [\fB\-p\fR|\fB\-\-port\fR <\fBport\fR>] [\fB\-\-version\fR] [\fB\-q\fR|\fB\-\-quiet\fR] [\fB\-?\fR|\fB\-\-help\fR] [\fB\-\-usage\fR] [\fB\-c\fR  \fB\-\-category\fR  \fBcategory\fR] [\fIfilename\fR \&.\&.\&.]

.SH "DESCRIPTION"

.PP
 \fIpilot\-install\-hinote\fR allows the user to write one or more files as a new \fIHi\-Note\fR or into the Hi\-Notes Palm application itself, on the Palm handheld\&.

.PP
Please see \fIhttp://www\&.cyclos\&.com/\fR for more information on Hi\-Note\&.

.SH "OPTIONS"

.SS "pilot-install-hinote option"

                        \fB\-c\fR,
                        \fB\-\-category\fR
                        <\fBcategory\fR>
                    
.PP
Write files to <\fBcategory\fR> in the Hi\-Note application itself\&.

.SS "Conduit Options"

                        \fB\-p\fR, \fB\-\-port\fR
                        <\fBport\fR>
                    
.PP
Use device file <\fIport\fR> to communicate with the Palm handheld\&. If this is not specified, \fIpilot\-install\-hinote\fR will look for the \fI $PILOTPORT \fR environment variable\&. If neither are found or supplied, \fIpilot\-install\-hinote\fR will print the usage information\&.

                        \fB\-q\fR, 
                        \fB\-\-quiet\fR
                    
.PP
Suppress 'Hit HotSync button' message

                        \fB\-v\fR, \fB\-\-version\fR
                    
.PP
Display version of \fIpilot\-install\-hinote\fR and exit without connecting\&.

.SS "Help Options"

                        \fB\-h\fR, \fB\-\-help\fR
                    
.PP
Display the help synopsis for \fIpilot\-install\-hinote\fR and exit without connecting\&.

                        \fB\-\-usage\fR 
                    
.PP
Display a brief usage message and exit without connecting\&.

.SH "USAGE"

.PP
The program will connect to a target Palm handheld, and write the file or files given on the commandline into the Hi\-Note application\&.

.SH "AUTHOR"

.PP
 \fIpilot\-install\-hinote\fR was originally written by Bill Goodman <bgoodman@cyclos\&.com>\&.

.PP
This manual page was converted to XML by Neil Williams <linux@codehelp\&.co\&.uk> 

.SH "REPORTING BUGS"

.PP
We have an online bug tracker\&. Using this is the only way to ensure that your bugs are recorded and that we can track them until they are resolved or closed\&. Reporting bugs via email, while easy, is not very useful in terms of accountability\&. Please point your browser to http://bugs\&.pilot\-link\&.org: \fIhttp://bugs.pilot-link.org\fR and report your bugs and issues there\&.

.SH "COPYRIGHT"

.PP
This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version\&.

.PP
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE\&. See the GNU General Public License for more details\&.

.PP
You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc\&., 51 Franklin St, Fifth Floor, Boston, MA 02110\-1301, USA\&.

.SH "SEE ALSO"

.PP
 \fIhinotes\fR(1), \fIpilot\-link\fR(7)\&.



src/pilot-install-hinote.c
/*
 * $Id: pilot-install-hinote.c,v 1.42 2006/10/17 13:24:07 desrod Exp $ 
 *
 * install-hinote.c:  Palm Hi-Note note installer
 *
 * Copyright 1997 Bill Goodman
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 *
 */

#include <stdio.h>
#include <sys/stat.h>

#include "pi-dlp.h"
#include "pi-hinote.h"
#include "pi-header.h"
#include "pi-userland.h"

/* Size of maximum buffer, containing 28k of note + a filename + \n + NUL at
   the end. */
#define HINOTE_BUFFER_SIZE (28 * 1024 + FILENAME_MAX + 2)

int main(int argc, const char *argv[])
{
	int 	db,
		sd	= -1,
		c,		/* switch */
		j,
		filenamelen,
		filelen,
		err,
		category 	= 0,
		note_size;

	const char
                *file_arg;

	char    *file_text,
		*cat 		= NULL;
	pi_buffer_t *appblock;

	unsigned char note_buf[0x8000];
	FILE 	*f;
	struct 	PilotUser User;
	struct 	stat info;
	struct 	HiNoteAppInfo mai;
	struct 	HiNoteNote note;

	poptContext pc;

	struct poptOption options[] = {
		USERLAND_RESERVED_OPTIONS
		{"category", 'c', POPT_ARG_STRING, &cat, 0, "Write files to <category> in the Hi-NOte application",
		 "category"},
		POPT_TABLEEND
	};

	pc = poptGetContext("install-hinote", argc, argv, options, 0);
	poptSetOtherOptionHelp(pc,"<file> ...\n\n"
		"   Install local files into your Hi-Notes database on your Palm device\n"
		"   Please see http://www.cyclos.com/ for more information on Hi-Note.\n\n"
		"   Example arguments:\n"
		"      -p /dev/pilot -c 1 ~/Palm/Note1.txt ~/Note2.txt\n\n");

	if (argc < 2) {
		poptPrintUsage(pc,stderr,0);
		return 1;
	}

	while ((c = poptGetNextOpt(pc)) >= 0) {
		fprintf(stderr,"   ERROR: Unhandled option %d.\n",c);
		return -1;
	}

	if (c < -1) {
		plu_badoption(pc,c);
	}

	if(poptPeekArg(pc) == NULL) {
		fprintf(stderr, "   WARNING: No files to install\n");
		return 0;
	}

	/* Allocate buffer for biggest possible file */
	file_text = (char *) malloc(HINOTE_BUFFER_SIZE);
	if (file_text == NULL) {
		fprintf(stderr,"   ERROR: Cannot allocate memory for files.\n");
		return 1;
	}


	sd = plu_connect();
        if (sd < 0)
                goto error;

        if (dlp_ReadUserInfo(sd, &User) < 0)
                goto error_close;

	/* Open Hi-Note's database, store access handle in db */
	if (dlp_OpenDB(sd, 0, 0x80 | 0x40, "Hi-NoteDB", &db) < 0) {
		fprintf(stderr,"   ERROR: Unable to open Hi-NoteDB on Palm.\n");
		dlp_AddSyncLogEntry(sd, "Unable to open Hi-NoteDB.\n");
		goto error_close;
	}

	appblock = pi_buffer_new(0xffff);
	j = dlp_ReadAppBlock(sd, db, 0, 0xffff, appblock);
	unpack_HiNoteAppInfo(&mai, appblock->data, j);	/* should check result */
	pi_buffer_free(appblock);

	if (cat != NULL) {
		category = plu_findcategory(&mai.category,cat,
			PLU_CAT_CASE_INSENSITIVE |
			PLU_CAT_MATCH_NUMBERS |
			PLU_CAT_DEFAULT_UNFILED);
	}

	while((file_arg = poptGetArg(pc)) != NULL) {

		/* Attempt to check the file size, stat() returns nonzero on
		   error */
		err = stat(file_arg, &info);
		if (err) {
		   /* FIXME: use perror() */
		   fprintf(stderr,"   WARNING: Checking file '%s': %s\n", file_arg,strerror(errno));
		   continue;
		}

		/* If size is good, open the file. */
		filelen = info.st_size;
		if (filelen > 28672) {

			fprintf(stderr,"   WARNING: This note (%i bytes) is larger than allowed size of 28k (28,672 bytes),\n"
			       "             please reduce into two or more pieces and sync each again.\n\n",
				filelen);

			continue;
		} else {
			f = fopen(file_arg, "r");
		}

		if (f == NULL) {
			fprintf(stderr,"   WARNING: Opening file '%s': %s\n",file_arg,strerror(errno));
			continue;
		}

		filenamelen = strlen(file_arg);


		memset(file_text,0,HINOTE_BUFFER_SIZE);
		strcpy(file_text, file_arg);
		file_text[filenamelen] = '\n';

		fread(file_text + filenamelen + 1, filelen, 1, f);
		file_text[filenamelen + 1 + filelen] = '\0';


		note.text = file_text;
		note.flags = 0x40;
		note.level = 0;
		note_size = pack_HiNoteNote(&note, note_buf, sizeof(note_buf));

		/* dlp_exec(sd, 0x26, 0x20, &db, 1, NULL, 0); */
		if (!plu_quiet) {
			fprintf(stdout, "   Installing %s to Hi-Note application...\n", file_arg);
		}
		dlp_WriteRecord(sd, db, 0, 0, category, note_buf,
				note_size, 0);
	}
	free(file_text);

	/* Close the database */
	dlp_CloseDB(sd, db);

	/* Tell the user who it is, with a different PC id. */
	User.lastSyncPC = 0x00010000;
	User.successfulSyncDate = time(NULL);
	User.lastSyncDate = User.successfulSyncDate;
	dlp_WriteUserInfo(sd, &User);

	dlp_AddSyncLogEntry(sd, "Successfully wrote Hi-Note notes to Palm.\n"
				"Thank you for using pilot-link.\n");
	dlp_EndOfSync(sd, 0);
	pi_close(sd);
	return 0;

error_close:
	pi_close(sd);

error:
	return -1;
}

/* vi: set ts=8 sw=4 sts=4 noexpandtab: cin */
/* ex: set tabstop=4 expandtab: */
/* Local Variables: */
/* indent-tabs-mode: t */
/* c-basic-offset: 8 */
/* End: */
